src/_pytest/pathlib.py

```diff
@@ -480,6 +480,7 @@ def import_path(
                 "Can't find module {} at location {}".format(module_name, str(path))
             )
         mod = importlib.util.module_from_spec(spec)
+        sys.modules[module_name] = mod
```

Comment:
Unfortunately AFAIK one of the "advantages" of using `importlib` was that it does not change `sys.path` or `sys.modules` at all... or perhaps the main advantage is actually not changing `sys.path` and updating `sys.modules` is fine?

@asottile what do you think? Also @The-Compiler which has some experience in this matter.

Reply:
I hope it's the latter because some code out there really depend on the module being in `sys.modules`, e.g., `ns = sys.modules.get(cls.__module__).__dict__` in `dataclasses.py`.

Reply:
I think we should strive to be as close as possible to a "normal" import in order to not break people's expectations. We definitely should avoid messing with `sys.path` (as this will change how other imports work!), but I think we should add our module to `sys.modules` so that our imports aren't exotic in any way.

The `importlib.import_module` convenience method also does that:

```pycon
>>> import importlib
>>> 'pytest' in sys.modules
False
>>> importlib.import_module('pytest')
<module 'pytest' from '/usr/lib/python3.8/site-packages/pytest/__init__.py'>
>>> sys.modules['pytest']
<module 'pytest' from '/usr/lib/python3.8/site-packages/pytest/__init__.py'>
```

Which makes me wonder - why are we hand-rolling everything here based on the more low-level functions rather than just using the convenience function? The importlib doc also has a ["Importing a source file directly"](https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly) recipe which looks very similar to what we do, and also adds the module to `sys.modules`.

Reply:
> I hope it's the latter because some code out there really depend on the module being in sys.modules, e.g., ns = sys.modules.get(cls.__module__).__dict__ in dataclasses.py.

Definitely! `importmode=importlib` definitely doesn't cover all cases, at least currently.

> If pytest really depend on modules not being added to sys.modules when using importlib, perhaps a test should be added for this behavior?

It is not that pytest depends on that, is how other modules interact with that and how that module is added to `sys.modules` (see below).

> Also, if it's the case, maybe the module could be added temporarily to sys.modules, just for spec.loader.exec_module(mod).

I had the same idea a few years back, but @RonnyPfannschmidt correctly pointed out that removing a module from `sys.modules` might cause havoc later (think local imports failing when the normal import succeeded).

> The importlib doc also has a "Importing a source file directly" recipe which looks very similar to what we do, and also adds the module to sys.modules.

Indeed. 

Just to recap, one of the reason this option exists is to allow different test modules with the same name, without a package, to be imported. This is the classic problem where the test hierarchy below fails:

```
tests/
  a/
    test_foo.py
  b/
    test_foo.py
```

(Without packages)

By the time we get to `b/test_foo.py`, under normal import mode, we change `sys.path` to point to `tests/b`, and try to import `test_foo.py`; but then we realize `test_foo` is already in `sys.modules`, but point to a different file, and raise a module mismatch error.

With `--importmode=importlib` the above works, because to import both `test_foo.py` files we don't change `sys.path` or `sys.modules`, so there's no conflict. But of course, having a module object which is not in `sys.modules` breaks people's expectations.

Now if we want to add both `test_foo` modules in `sys.modules`, which key should we use to avoid conflicts? Note that we can't rely on `sys.path` at all (as we don't want to change it). Perhaps use the full file path, replacing `os.path.sep` by `.`, would resolve this?

Full disclosure: I'm just stating how things work and the rationale behind it. I'm more than happy to improve this. 😁

Reply:
- This brings back the topic of import conflicts and requires some discussion and crosscheck that i can't do soon

Reply:
I'm late to the party sorry was very busy today :)

the points above are what make this tricky: the `importlib` mode was designed to avoid module namespace conflicts and the keys in `sys.modules` would be those same conflicts (so we'd need to come up with some unique thing that *also* gets set as `__name__` in that module)

Reply:
The general idea i have for that is a general concept of collection roots that either belong to a existing package, or get a otherwise unique name

Unfortunately i don't have the capacity to work on this soon

Reply:
How about using the module name relative to `config.root`, perhaps adding a uniquely-named prefix? I suppose that would produce unique names and also something one would look and recognize.

So in the usual project:

```
src/
tests/
pytest.ini
```

We can use `__pytest_importmode_importlib__.tests.a.test_foo`, I doubt this would clash with anything (just using `tests.a.test_foo` was my initial idea, but this can clash depending on the project layout and `config.root`).

I believe this module name can be considered experimental/internal detail (and warn that in the docs as such) until we are sure it works or come up with a better idea.

Reply:
I've done some manual testing (comparing pytest against importing things manually, also testing running pytest with multiple roots). Note that even though some modules with the same name won't appear in `sys.modules`, everything already seems to work fine (with the changes in this PR), even with `conftest` files. (Maybe there's some overlooked use case?)

It would probably be problematic if tests tried to import other test files, but I think that they are not supposed to do that, specially with implicit namespace packages.

The tests can be found here: https://github.com/tadeu/_pytest-7870

Diff between running "manually" and with pytest is here: https://github.com/tadeu/_pytest-7870/blob/main/diff1.diff

I'd be happy to transform that into unit tests in pytest/testing if you think it's worthy.

Reply:
Still overwriting the modules will definitely prove problematic in real world test suites... pytest is used by thousands of projects, I'm sure this will break some people.

I think the proper way to handle this is to use a unique key to insert the module in `sys.modules`.

Reply:
> We can use `__pytest_importmode_importlib__.tests.a.test_foo`, I doubt this would clash with anything (using only `tests.a.test_foo` was my initial idea, but this can clash depending on the project layout and `config.root`).

After the discussions I can also try to implement this, I hope it's not too much of a rabbit hole in the internals 😁🐰

Reply:
Thanks! Let's see what others think first though. 👍

Reply:
If that works with dataclasses, pickle, or the other scenarios mentioned in #7245, I'd say go for it? I think we don't necessarily need to get *everything* right as part of this PR. It's clearly an improvement of the status quo (where pytest breaks with a variety of stdlib/builtin functionality).

Given that `--import-mode=importlib` isn't the default yet and we're asking for feedback with it, I think it's okay to have slightly breaking changes in the future (before it's made the default) rather than stalling this PR further because we don't actually know what's the right approach.

Reply:
I agree with @The-Compiler's reasoning.

What others think? Can we give green light to @tadeu to implement the idea of using a unique key based on `config.root`? Other ideas?

Reply:
Anybody? It would be interesting to get this kicking again.